documentation/c10-classdesign.sil

```diff
@@ -1,20 +1,301 @@
+  -- Our own commands come here
+end
+\end{raw}
+
+Within it, use the \code{self:registerCommand} function.
+It takes three parameters: a command name, a function to implement the command, and some help text.
+
+The signature of a function representing a SILE command is fixed:
+you need to take two parameters, \code{options} and \code{content}.\footnote{%
+Of course ou can name your parameters whatever you like, but these are the most common names.}
+Both of these parameters are Lua tables.
+The \code{options} parameter contains the command’s parameters as a key-value table, and the \code{content} parameter is an abstract syntax tree reflecting the input being currently processed.
+
+So in the case of \autodoc:command[check=false]{\mycommand[size=12pt]{Hello \break world}},
+the first parameter will contain the table \code{\{size = "12pt"\}} and the second parameter will contain the table:
+
+\begin[type=autodoc:codeblock]{raw}
+{
+  "Hello ",
+  {
+    options = {},
+    id = "command",
+    pos = …,
+    col = …,
+    lno = …,
+    command = "break"
+  },
+  " world"
+}
+\end{raw}
+
+Most commands will find themselves doing something with the \code{options} and/or calling \hbox{\code{SILE.process(content)}} to recursively process and render the argument.
+
+Here’s a very simple example: a \autodoc:command[check=false]{\link} command may take an \code{href} attribute.
+We want to render \autodoc:command[check=false]{\link[href=http://...]{Hello}} as \examplefont{Hello (\code{http://...})}.
+First we need to render the content, and then we need to do something with the attribute.
+We use the \code{SILE.typesetter:typeset} and \code{SILE.call} functions to output text and call other commands.
+
+\begin[type=autodoc:codeblock]{raw}
+  self:registerCommand("link", function(options, content)
+    SILE.process(content)
+    if (options.href) then
+      SILE.typesetter:typeset(" (")
+      SILE.call("code", {}, { options.href })
+      SILE.typesetter:typeset(")")
+    end
+  end)
+\end{raw}
+
+Now, let’s design a (naive) \autodoc:environment[check=false]{blockquote} environment implementing indented (and possibly nested) quotations.
+You do remember, right, that an environment in SILE is not much different from a command?
+So a command be it, without any option this time, but playing with vertical skip, measurements, glue, (temporary) left and right margin settings.
+(If these concepts elude you, consider re-reading the previous chapters where they are introduced.)
+
+\begin[type=autodoc:codeblock]{raw}
+  self:registerCommand("blockquote", function (_, content)
+    SILE.call("smallskip")
+    SILE.settings:temporarily(function ()
+      local indent = SILE.measurement("2em"):absolute()
+      local lskip = SILE.settings:get("document.lskip") or SILE.nodefactory.glue()
+      local rskip = SILE.settings:get("document.rskip") or SILE.nodefactory.glue()
+      SILE.settings:set("document.lskip",
+        SILE.nodefactory.glue(lskip.width + indent))
```

Comment:
Erm. It works in Lua 5.4 (tested separately), but `lskip.width` is a length, `indent` a measurement. Mixing apples and oranges again?